### PR TITLE
[OpenGL] [sparse] Basic support for dynamic SNode

### PR DESCRIPTION
**Pull request by archibate**
Related issue = #711

[[Click here for the format server]](http://kun.csail.mit.edu:31415/)

----
Basic support!

1. Use `for i in range(dynamic_len): list[i] = i` for listgen (low efficiency but works).
2. No non-top dynamic snode like `ti.root.dense.dynamic` (non-trivial, will do iapr).

----
I also noticed that making msgbuf[] size to be explicit cause NV GLSL compiler to be slow. Should I open an extra buf for msg to make NV quicker? It may push mesa GL more close to it's buf number limit of 12.

### CODE REVIEW

**Comment by archibate** [comment]
Hello? I thought I made a big step..

**Comment by archibate** [comment]
Hello? 7 days passed and nobody react me except bots.

File: taichi/backends/opengl/codegen_opengl.cpp [hunk 1]
```diff
@@ -616,6 +699,103 @@ class KernelGen : public IRVisitor {
...
+    emit("_list_len_ = _data_i32_[{} >> 2];", addr);
+    emit("for (int i = 0; i < _list_len_; i++) {{");
+    {
+      ScopedIndent _s(line_appender_);
+      emit("_list_[i] = i;");
+    }
+    emit("}}");
+  }
+
+  void generate_listgen_for_dense(const SNode *snode) {
```
Comment (k-ye) [inline taichi/backends/opengl/codegen_opengl.cpp:766]: It feels a bit strange: if we use this on a dense with more than one dimension, it probably won't work for now. But struct-for on a dense-only hierarchy is demoted to range-for, so it doesn't matter..

File: taichi/backends/opengl/codegen_opengl.cpp [hunk 2]
```diff
@@ -291,6 +301,77 @@ class KernelGen : public IRVisitor {
...
+             get_snode_meta_address(stmt->snode),
+             stmt->input_index->short_name());
+      } else {
+        TI_NOT_IMPLEMENTED
+      }
+    }
+  }
+
+  void visit(SNodeOpStmt *stmt) override {  // IAPR?
+    if (stmt->op_type == SNodeOpType::activate) {
```
Comment (k-ye) [inline taichi/backends/opengl/codegen_opengl.cpp:318]: FYI, I don't think `SNodeOpType::activate` will be directly used. E.g. https://github.com/taichi-dev/taichi/blob/37976e750e1579ff9328f3d25f1a63cb3b3292e2/taichi/codegen/codegen_llvm.cpp#L904-L925 There isn't a handling for `::activate`.

IIUC, `::activate` is always translated into an `SNodeLookupStmt` to decide whether this is a strong or weak lookup. But I feel like it's fine to leave it here. Maybe we will have something like `ti.activate(x, [i, j, k])` one day? @yuanming-hu

File: taichi/backends/opengl/codegen_opengl.cpp (hunk 1 above)
Comment (archibate) [reply taichi/backends/opengl/codegen_opengl.cpp:766]: Not sure how to support multi-dim listgen now, will do iapr.

**Comment by archibate** [comment]
> LGTM! Sorry, back then I was wondering whether it's good to support a somewhat limited `dynamic`. I'd also like to hear @yuanming-hu 's opinion on this. Can we discuss on this kind of decision before implementing it next time?

The limited support here is good enough for ordinary purpose. Taichi THREE for example, only wants extend-able array for loading meshes.

File: taichi/backends/opengl/codegen_opengl.cpp (hunk 2 above)
Comment (archibate) [reply taichi/backends/opengl/codegen_opengl.cpp:318]: Great, feel free to open an issue in `ti.activate` and assign me later.

**Comment by archibate** [comment]
@yuanming-hu Could you take a pass? It's becoming stale.

File: taichi/backends/opengl/codegen_opengl.cpp (hunk 2 above)
Comment (yuanming-hu) [reply taichi/backends/opengl/codegen_opengl.cpp:318]: Thanks for the discussions. I think having `ti.activate` is a good idea.

**Pull request merged**